apps/prairielearn/src/lib/course.ts

```diff
@@ -142,6 +142,9 @@ export async function pullAndUpdateCourse({
           job.info('Clean local files not in remote git repository');
           await job.exec('git', ['clean', '-fdx'], gitOptions);
 
+          job.info('Check out current branch');
+          await job.exec('git', ['checkout', branch], gitOptions);
```

Comment:
I want to think through the sequencing here. Is there a concern with doing the `git clean` before `git checkout`? I'm trying to imagine some sequence that would leave undesired files on disk; maybe the fact that I can't immediately think of one means we're fine? Or maybe it's safer to move the `checkout` before the `clean`? I'm not 100% sure.

Reply:
My thinking is that it's best to do the checkout after the clean, as you have it now. My thinking is that we want to make sure the checkout succeeds, so we don't want it to be blocked by having a spurious untracked file on disk that might conflict with a file from the new checkout. After the clean, we'll be in a known-good state, so the following checkout should succeed with no problems.

If we do the checkout before the clean then we run the risk that some untracked file will block the checkout.

Reply:
I also wonder whether we should maybe do a `git restore --staged --worktree` _before_ the clean, just to make extra sure that we've got state on disk that we expect?

Reply:
Sounds good, we'll keep the ordering of `clean` and `checkout` as they are now. Re: `git restore ...`, is there a particular bad case you're thinking of that this would solve, or are you just being extra cautious?

Reply:
This is definitely just in the realm of being extra cautious. I guess I was thinking "if I have a git repo in an unknown state, how do I get it back to a known safe state". That is, I'm assuming that there might be some partially-edited files locally, maybe even some files staged for commit, and some other junk lying around in the current directory. So I think that:
* `git restore --staged --worktree` should fix up all files that git is tracking
* `git clean -fdx` should eliminate any files that git is not tracking
* `git checkout ...` will then move us to the correct branch

Reply:
I'm ok with being extra cautious here! Are you thinking we'd want this in the in-browser-editor flows as well?

Reply:
I think for now I'd be fine with just doing on it sync, because that at least gives a way to reset everything back to a safe state. I'd be open to doing this before every git operation, but I'm a bit unsure about how `checkout` would work because we might not have the correct branch to checkout yet (if it's only in the remote repo).

Reply:
I added `git restore ...` in 9df491c.